tests/test_regression/test_losses.py

```diff
@@ -0,0 +1,140 @@
+        )
+        return instance.process_lcwa_scores(
+            predictions=predictions, labels=labels, label_smoothing=self.label_smoothing, num_entities=self.num_entities
+        )
+
+
+@dataclasses.dataclass
+class SLCWATestCase(LossTestCase):
+    """Test SLCWA scores."""
+
+    batch_size: int
+    label_smoothing: float | None
+    num_entities: int
+    num_negatives: int
+
+    # docstr-coverage: inherited
+    def __call__(self, instance: Loss, generator: torch.Generator) -> torch.Tensor:
+        positive_scores = torch.rand(self.batch_size, 1, generator=generator)
+        negative_scores = torch.rand(self.batch_size, self.num_negatives, generator=generator)
+        return instance.process_slcwa_scores(
+            positive_scores=positive_scores,
+            negative_scores=negative_scores,
+            label_smoothing=self.label_smoothing,
+            batch_filter=None,
+            num_entities=self.num_entities,
+        )
+
+
+test_case_resolver = Resolver(classes=[LCWATestCase, SLCWATestCase], base=LossTestCase, suffix="testcase")
+
+
+def get_cases() -> list[tuple[Loss, LossTestCase, float]]:
+    """Get loss test cases."""
+    path = pathlib.Path(__file__).parent.joinpath("data", "losses.json")
+    rv = []
+    for data in json.loads(path.read_text()):
+        loss = loss_resolver.make(data["loss"])
+        test_case: LossTestCase = test_case_resolver.make(data["type"], data["kwargs"])
+        rv.append((loss, test_case, data["value"]))
+    return rv
+
+
+def _ids(x: type | LossTestCase) -> str:
+    """Determine part of test case name."""
+    if isinstance(x, LossTestCase):
+        return str(x)
+    return x.__name__
+
+
+@pytest.mark.parametrize(("instance", "case", "expected"), get_cases())
+def test_regression_2(instance: Loss, case: LossTestCase, expected: float, generator: torch.Generator) -> None:
+    """Check whether the loss value is the expected one."""
+    actual = case(instance=instance, generator=generator)
+    assert torch.isclose(torch.as_tensor(expected), actual)
+
+
+@pytest.mark.parametrize(
+    ("cls", "case"),
+    itertools.product(
+        set(loss_resolver),
+        [
+            LCWATestCase(batch_size=1, label_smoothing=None, num_entities=32),
+            SLCWATestCase(batch_size=1, label_smoothing=None, num_entities=32, num_negatives=3),
+        ],
+    ),
+    ids=_ids,
+)
+def test_regression(cls: type[Loss], case: LossTestCase, generator: torch.Generator) -> None:
+    """Check whether the loss value is the expected one."""
+    # create instance with default parameters.
+    instance = loss_resolver.make(cls)
+
+    # get loss value for the given case
+    loss_value = case(instance=instance, generator=generator)
+
+    # determine reference file path
+    name = loss_resolver.normalize_cls(cls)
+    path = DATA_ROOT.joinpath(name).with_suffix(".json")
+
+    # TODO: is there a nicer way how to enable generating missing values more explicitly?
```

Comment:
@cthoyt Do you have an idea for this situation? Basically, I want to have an easy way to generate missing reference files / explicitly update one in case of breaking changes.

One way would be to move the code into the package (i.e. out of `tests/`), but the code is only needed for these tests, so it does not feel right in the main package.

Reply:
I was just hacking on it. Might have an idea attached to the JSON update I just made, but it might be more lines of code 😁

Reply:
Done in 65bb2416